adr/ADR-33.md

```diff
@@ -0,0 +1,251 @@
+letters. Prefixed by `$SRV`. Note that this prefix needs to be overridable much
+in the way as we do for `$JS`, in order to enable targetting tools to work across
+accounts.
+
+The initial _verbs_ supported by the service include:
+
+- `PING`
+- `STATS`
+- `INFO`
+- `SCHEMA`
+
+Using the above verbs, it becomes possible to build a service subject hierarchy
+like:
+
+`$SRV.PING|STATS|INFO|SCHEMA` - pings and retrieves status for all services
+`$SRV.PING|STATS|INFO|SCHEMA.<name>` - pings or retrieves status for all
+services having the specified name `$SRV.PING|STATS|INFO|SCHEMA.<name>.<id>` -
+pings or retrieves status of a particular service instance
+
+Services should respond to:
+
+- All service requests
+- All service requests that match their `name`
+- All services requests that match their `name` and `id`
+
+### INFO
+
+Returns a JSON having the following structure:
+
+```typescript
+{
+    /**
+     * The kind of the service reporting the status
+     */
+    name: string,
+    /**
+     * The unique ID of the service reporting the status
+     */
+    id: string,
+    /**
+     * Description for the service
+     */
+    description: string,
+    /**
+     * Version of the service
+     */
+    version: string,
+    /**
+     * Subject where the service can be invoked
+     */
+    subject: string
+}
+```
+
+All the fields above map 1-1 to the metadata provided when the service was
+created. Note that `subject` is the subject that the service is listening for
+requests on.
+
+### PING
+
+Returns the following schema:
+
+```typescript
+{
+    /**
+    * The kind of the service reporting the status
+    */
+    name: string,
+    /**
+    * The unique ID of the service reporting the status
+    */
+    id: string,
+}
+```
+The intention of `PING` is for clients to calculate RTT to a service.
+
+### SCHEMA
+
+Returns a JSON having the following structure. Note that the `schema` struct is
+only returned if the `schema` was specified when created.
+
+```typescript
+{
+    /**
+     * The kind of the service reporting the status
+     */
+    name: string,
+    /**
+     * The unique ID of the service reporting the status
+     */
+    id: string,
+    /**
+     * Version of the service
+     */
+    version: string,
+    /**
+     * The schema specified when the service was created
+     */
+    schema?: {
+        /**
+         * A string or URL
+         */
+        request: string,
+        /**
+         * A string or URL
+         */
+        response: string
+    };
+}
+```
+
+### STATS
+
+The STATS request can have an optional JSON payload `{ internal: boolean }`. If
+specified and `true` the service should report all its internal endpoints, thus
+the reson the stats returns an array of EndpointStats:
+
+```typescript
+{
+  name: string;
+  id: string;
+  version: string;
+  stats: EndpointStats[];
+}
+```
+
+EndpointStats looks like:
+
+```typescript
+{
+    /**
+     * Name of the endpoint
+     */
+    name: string;
```

Comment:
On endpoint stats for discovery subjects, is the EndpointStats.name like this `$SRV.INFO` or just `INFO`. Maybe add this in the comment for the field